fuel/datasets/hdf5.py

```diff
@@ -64,3 +67,124 @@ def get_data(self, state=None, request=None):
+      point (exclusive) of the split.
+
+    The `which_set`, `start` and `stop` parameters work together in the
+    following way:
+
+    * `which_set` is resolved first. If it is `None`, the whole dataset is
+      used.
+    * `start` and `stop` define a slice *within the context of*
+      `which_set`.
+
+    Parameters
+    ----------
+    path : str
+        Path to the HDF5 file.
+    which_set : str, optional
+        Name of the root group attribute containing the split information.
+        Defaults to `None`, in which case the whole dataset is used.
+    subset : slice, optional
+        A slice of data *within the context of the split* to use. Defaults
+        to `None`, in which case the whole split is used. **Note:
+        at the moment, `slice.step` must be either 1 or `None`.**
+    load_in_memory : bool, optional
+        Whether to load the data in main memory. Defaults to `False`.
+    driver : str, optional
+        Low-level driver to use. Defaults to `None`. See h5py
+        documentation for a complete list of available options.
+
+    """
+    ref_counts = dict()
+
+    def __init__(self, path, which_set=None, subset=None, load_in_memory=False,
+                 driver=None, **kwargs):
+        self.path = path
+        self.which_set = which_set
+        self.subset = subset if subset else slice(None, None, None)
+        self.load_in_memory = load_in_memory
+        self.driver = driver
+        self.load()
+        super(H5PYDataset, self).__init__(**kwargs)
+
+    def _get_file_id(self):
+        try:
+            return next(
+                filter(lambda x: x.name == self.path, self.ref_counts.keys()))
+        except StopIteration:
+            return self.path
+
+    def load(self):
+        handle = self._out_of_memory_open()
+        self.provides_sources = list(handle.keys())
+        shapes = [data_source.shape for data_source in handle.values()]
+        if any(s[0] != shapes[0][0] for s in shapes):
+            raise ValueError("sources have different lengths")
+        if self.subset.step not in (1, None):
+            raise ValueError("subset.step must be either 1 or None")
+        start, stop = (handle.attrs[self.which_set] if self.which_set
+                       else (0, shapes[0][0]))
+        self.subset = slice(
+            start if self.subset.start is None else self.subset.start,
+            stop if self.subset.stop is None else self.subset.stop,
+            self.subset.step)
+        self.num_examples = self.subset.stop - self.subset.start
+        self.data_sources = ([data_source[self.subset] for
+                              data_source in handle.values()]
+                             if self.load_in_memory else None)
+        self._out_of_memory_close(handle)
+
+    def open(self):
+        return None if self.load_in_memory else self._out_of_memory_open()
+
+    def _out_of_memory_open(self):
+        file_id = self._get_file_id()
```

Comment:
I'm a bit confused here, but maybe that's just because I don't know `h5py`'s interface very well. `file_id` can be either a path (string) or a file handle here, right? But then you pass the `file_id` to `h5py.File`, does that make sense?